feat: add `queryInterface.changeColumns`, fix many bugs with `changeColumn`

ephys opened the pull request:
## Description Of Change

The initial goal of this PR was to investigate [a bug reported here](https://github.com/sequelize/sequelize/issues/9058#issuecomment-1165615540), it quickly turned out that `changeColumn` had a completely broken implementation and I opted for doing a complete rewrite and fixing many bugs in the process.

### `changeColumns`

I've added `changeColumns` which is able to change more than one column from the same table in one `ALTER TABLE` query, reducing the number of necessary queries.

`changeColumns` follows our new API design, meaning it's possible to use it with a Model, a table name, a table object (table+schema), and it will default to the schema defined in the sequelize options if none is specified.

That method also throws clear error messages when the user tries to do something out of its scope. Previously, it would just crash with an obscure message.

`changeColumn` just defers to `changeColumns`, and inherits from its features.

### All dialects

- `changeColumns` *only* changes what has been provided. No more dropping the default value because it was not provided in the call. Instead `dropDefaultValue` has been added as an option. Similarly, the nullability of the column is *only* changed if `allowNull` is explicitly set to true or false (undefined just does not change the definition at all).
- The previous implementation was broken because it used regexes to modify another query, which broke very easily by using keywords in the default value.
- This method supports enabling/disabling `autoIncrement` & `autoIncrementIdentity`

### Postgres

- `changeColumns` is able to replace an enum with another, drastically simplifying the modification of enums in migrations. Before, it was impossible to use `changeColumn` to change the definition of a column's enum.

## Commit message

```

BREAKING CHANGE: `changeColumn` does not change a property if it is not specified. For instance, previously it would default `allowNull` to true and add `DROP NOT NULL` if unspecified.
postgres: `changeColumn` can replace a column's enum, as long as only that column is using that enum (which is the default behavior for tables generated by Sequelize).
`changeColumn` does not break if a SQL keyword is used in the default message anymore.
```

## Todos

- [x] postgres
- [ ] db2
- [ ] ibmi
- [ ] mariadb
- [ ] mssql
- [ ] mysql
- [ ] snowflake
- [ ] sqlite

WikiRik commented:
Does this replace #14477?

ephys commented:
Yep! It does fix the issue I mentioned but still checks whether a partially built query, that can include user-provided values (through `defaultValue`) contains some keywords, meaning the query will break if the `defaultValue` option includes one of those keywords

I'll merge the necessary changes from that PR into this one

ephys commented:
So I don't forget: This PR is blocked, we need to open another PR to clean up `queryInterface.describeTable` to ensure it returns consistent results accros all dialects

WikiRik commented:
Just to check; this is still blocked by `describeTable`, correct?

Any chance parts can be done already? Like migrating the enum functions to TS.

ephys commented:
Yep! I already introduced the internal query generator files & other similar refactors in a recent PR, the `canHaveDefaultValue` logic could be moved too, as well as the changes to `withUsageContext`, `getForeignKeyReferencesForTable`. The change to `Nullish` can also be done. Maybe the enum functions but I'm not sure. At least some of them can. Accepting `DbObjectId` in `Identifier` and other places would be great too, at `sql.identifier` could then directly specify the schema. Not using `.field` or `.fieldName` in our source code to remove deprecation warnings would be great too.

I added `MapView#getOrThrow` in #17077

Those would already be pretty great. As for the rest, it's indeed mostly blocked by `describeTable`

ephys commented:
(ignore notification, was demoing something)

lohart13 commented:
Just wondering what the status of this PR is?

I notice the last comments mention the issues with `describeTable` but was wondering if some of the changes such as the datatype ones (especially the ENUM's for Postgres) extracted into a separate PR?

I'm working on a `addColumn` rewrite some of the changes in the PR would greatly help.

WikiRik commented:
> Just wondering what the status of this PR is?

It's on hold. I think this will land somewhere later in v7 as an experimental feature and becomes the default in v8.

> I notice the last comments mention the issues with `describeTable` but was wondering if some of the changes such as the datatype ones (especially the ENUM's for Postgres) extracted into a separate PR?

As far as I know, those have not been extracted yet but feel free to do so.
